### PR TITLE
Detekt

### PR DESCRIPTION
**Pull request by pocmo**
Recently I disabled "findbugs" for Kotlin source files. There were too many false positives that were just not actionable and confusing; because the Kotlin code was nothing like the Java code findbugs assumed from the byte code.

However having no static code analyzer for Kotlin was a bit scary so I looked into getting [detekt](https://github.com/arturbosch/detekt) running.

I used the default detekt configuration with the following modifications:

* **CommentOverPrivateFunction**: False
  According to the rule comments for private functions should be avoided in favor of functions with expressive names. Currently we do have some private functions with comments and I do not think a comment explaining reasoning is harmful.

* **CommentOverPrivateProperty**: False
  See above. We currently have such comments and I don't think that's harmful.

* **ReturnCount** max: 2 -> 3
  We have been using early returns in the past to segment code and make it more readable. Having too many returns might be confusing and hurt readability. With 3 we cover our existing code. So I opted for just raising this limit.

* **WildcardImport**: False
  Detekt doesn't like wildcard imports by default and with that is following Java code style conventions. In general I agree but unfortunately by default Android Studio / IntelliJ will use wildcard imports for 5 or more imports from the same package (only in Kotlin code, not Java). Fighting the default setting and forcing all developers to re-configure their IDE seems excessive. So I just disabled this rule for now. However it is noteworthy that Google released official Kotlin Style Guidelines for Android recently. And they do not allow wild card imports:
  https://android.github.io/kotlin-guides/style.html#import-statements

### CODE REVIEW

**Comment by mcomella** [comment]
@pocmo How did you decide on detekt? Are there other static analysis tools available and if so, what are the trade-offs between them (assuming we intend to use detekt over them, rather than using them both)?

**Comment by mcomella** [comment]
First pain point: there is no online documentation for rules and why they fail. I have to resort to reading the rules source during code review (e.g. [`ComplexMethod`](https://github.com/arturbosch/detekt/blob/30b8b0945064889f00e1552d55868ec7c05de4d5/detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/ComplexMethod.kt#L21)) or pulling down the broken source and running it locally.

Note that there's a bug open to improve this but it hasn't been touched in a month: https://github.com/arturbosch/detekt/issues/496

**Comment by mcomella** [comment]
> In general I agree but unfortunately by default Android Studio / IntelliJ will use wildcard imports for 5 or more imports from the same package (only in Kotlin code, not Java). Fighting the default setting and forcing all developers to re-configure their IDE seems excessive. 

iirc, I think Java, by default, does wildcard imports (though maybe they changed that default setting in AS rather than IJ).

File: app/src/main/java/org/mozilla/focus/browser/DisplayToolbar.kt [hunk 1]
```diff
@@ -26,6 +26,7 @@ class DisplayToolbar(
         addOnOffsetChangedListener(this)
     }
 
+    @Suppress("MagicNumber")
```
Comment (mcomella) [inline app/src/main/java/org/mozilla/focus/browser/DisplayToolbar.kt]: nit: add a brief comment explaining why we chose this

**Comment by pocmo** [comment]
> @pocmo How did you decide on detekt? Are there other static analysis tools available and if so, what are the trade-offs between them (assuming we intend to use detekt over them, rather than using them both)?

I tried detekt and ktlint. detekt seems to:
* Cover correctness and style. While ktlint seems to focus more on code style.
* More active with more contributors
* Favored by the twitterverse 🙃 https://twitter.com/Anti_Hype/status/935140558711947264
* Easy to integrate and run.

Nevertheless I want to see how much overlap there is and whether running ktlint additionally makes sense.

> First pain point: there is no online documentation for rules and why they fail. I have to resort to reading the rules source during code review (e.g. ComplexMethod) or pulling down the broken source and running it locally.

Yep, that's true and was annoying for me as well. I often googled "detekt ErrorType" to understand the problem. I was thinking to pick up the HTML report issue myself in my free time. Anyhow, awkward reports (that require additional search) are still better than none at all - right?

File: telemetry [hunk 2]
```diff
@@ -0,0 +1 @@
+../telemetry-android/telemetry
```
[the remark this answers is not in the document]
Comment (pocmo) [reply telemetry]: Argh. Switching between branches and committing too much.. At some point I wondered where my soft link went. Apparently into this commit. :)

File: build.gradle [hunk 3]
```diff
@@ -21,6 +22,19 @@ buildscript {
...
 }
 
+plugins {
+    id "io.gitlab.arturbosch.detekt" version "1.0.0.RC5-5"
+}
+
+detekt {
+    version = "1.0.0.RC5-5"
+    profile("main") {
+        input = "$projectDir/app"
```
[the remark this answers is not in the document]
Comment (pocmo) [reply build.gradle]: Yeah. This is from the example in the detekt README. Running it over $projectDir seems to work fine; I'm changing this.

**Comment by arturbosch** [comment]
Hey, cool that your trying out detekt! Just some comments:

You do not have to use `@Suppress("MagicNumber")` to disable the reporting and pollute your code. You can also use a [baseline](https://github.com/arturbosch/detekt#code-smell-baseline-and-ignore-list).
The gradle plugin can generate yourself a baseline file with `gradle detektBaseline`. This will also add all your findings to it and they won't get reported next time.

We are working on more documentation and rule descriptions and reasoning behind them. Marvin is working on [generating docs](https://github.com/arturbosch/detekt/tree/master/detekt-generator/documentation) based on kdoc which will be hosted on a website in the future. Yes there is lot of kdoc missing but every rule should have a `issue.description` which is printed when using the `output` parameter.

**Comment by pocmo** [comment]
@arturbosch Thank you so much for commenting here. :)

> You do not have to use @Suppress("MagicNumber") to disable the reporting and pollute your code. You can also use a baseline.

Ah, interesting! There are some things reported currently that I still want to fix though. :) Ideally I want warningThreshold and failThreshold both set to 0 at some point. What I like about suppress is that I can add a reasoning why I want to suppress something. I guess with baseline this is not possible?

> We are working on more documentation and rule descriptions and reasoning behind them. Marvin is working on generating docs based on kdoc which will be hosted on a website in the future. Yes there is lot of kdoc missing but every rule should have a issue.description which is printed when using the output parameter.

Cool! Maybe we can help with that a bit.

@arturbosch While you are here ... You probably know best if it makes sense to use detekt with ktlint or if the coverage overlaps anyways. I saw a ktlint wrapper being mentioned and that some style rules will be removed from detekt?

**Comment by arturbosch** [comment]
Well as the baseline is just a xml file ofc you can place comments there too. If you like to know or let others know that yes this class is huge or have other issues but 'this' is the reason for it and this should stay ofc use `@Suppress` I would say. If the file is like never touched just throw it in the baseline.
The output from the baseline will not be printed in the console though.

Documentation fixes and additions are always welcome! 

Yes it makes sense to use both tools. KtLint for formatting and detekt for complexity or other issues (UnsafeCast come handy a few times!). Before RC5 we had all rules KtLint had, but as it seems KtLint gets more active now, we do not want to double spend the effort on formatting. It is fairly easy to wrap all ktlint rules in detekt and get all fixes from there asap. There might be problems using both tools as we both use the kotlin compiler (maybe in different versions) as I remember from this [issue](https://github.com/shyiko/ktlint/issues/22). However on our side we hadn't such reportings and I know some guys using both tools together.

**Pull request merged**